### PR TITLE
ignore comments when finding the start of pattern

### PR DESCRIPTION
**Pull request by kngwyu**
For #816 and #818.

### CODE REVIEW

File: src/racer/util.rs [hunk 1]
```diff
@@ -23,6 +27,114 @@ pub fn is_ident_char(c: char) -> bool {
...
+    }
+}
+
+impl Default for CommentAutomatonState {
+    fn default() -> CommentAutomatonState {
+        CommentAutomatonState::None
+    }
+}
+
+#[derive(Debug, Default)]
```
Comment (TedDriggs) [inline src/racer/util.rs]: Please add a rustdoc comment explaining what this is for.

File: src/racer/util.rs [hunk 2]
```diff
@@ -23,6 +27,114 @@ pub fn is_ident_char(c: char) -> bool {
...
+pub struct CommentAutomaton {
+    state: CommentAutomatonState,
+}
+
+impl CommentAutomaton {
+    pub fn cur_state(&self) -> bool {
```
Comment (TedDriggs) [inline src/racer/util.rs]: Please rename to `is_in_comment`

File: src/racer/util.rs [hunk 3]
```diff
@@ -23,6 +27,114 @@ pub fn is_ident_char(c: char) -> bool {
...
+enum CommentAutomatonState {
+    None,
+    Start, // just / occurs
+    InComment, // /* ...
+    BeforeEnd, // /* ... *
+    InSlash, // like ///// (maybe not useless)
+}
+
+impl CommentAutomatonState {
+    // in case we can know wer'e in comment without looking ahead
```
Comment (TedDriggs) [inline src/racer/util.rs]: I don't understand this comment; or maybe it's that the method name doesn't match the comment?

Typo: Should be "we're"

File: src/racer/util.rs [hunk 4]
```diff
@@ -23,6 +27,114 @@ pub fn is_ident_char(c: char) -> bool {
...
 }
 
+// to ignore C-Style Comment
+#[derive(Debug)]
+enum CommentAutomatonState {
+    None,
+    Start, // just / occurs
+    InComment, // /* ...
+    BeforeEnd, // /* ... *
+    InSlash, // like ///// (maybe not useless)
```
Comment (TedDriggs) [inline src/racer/util.rs]: Please use rustdoc comments for each enum variant, rather than postfixed comments.

**Comment by kngwyu** [comment]
Thanks for the review, and I came up with a reimplemention of  ```CommentAutomaton```  to iterator, such as ```CommentSkipIterRev``` for easiness to read.
Do you think it's better?

**Comment by kngwyu** [comment]
Updated(packed Comment-Skip logic into ```CommentSkipIterRev``` in codecleaner.rs and add some tests).
And the test which fails isn't related to my commits so I created another PR(#820).

**Comment by TedDriggs** [comment]
> In addition, I used rustfmt to system.rs by mistake... (Should I fix this?)

Yes please. `rustfmt` is good, but it'll make this much harder to review. FYI, I'm going to try and review this early next week.

**Comment by kngwyu** [comment]
Thanks for advice.
I fixed and rebased.

**Pull request merged**